book/book.toml

```diff
@@ -14,3 +14,4 @@ command = "mdbook-mermaid"
 
 [output.html]
 additional-js = ["mermaid.min.js", "mermaid-init.js"]
+additional-css = ["theme/css/custom.css"]
```

Comment:
I really like the diagram!

I think one of the dependencies might be inverted:
- all the arrows from the inbound service should go in the same direction, because nothing depends on it

Optional: I think it might be easier to read if we:
- switch how dependencies are shown, so the called service is the target of the arrow
- split the block and checkpoint verifier out of the chain verifier (they have different dependencies, the checkpoint verifier doesn't verify transactions)

What do you think?

Reply:
Also, we're missing:
- the transparent script, shielded signature, and shielded proof verifiers (which I think is fine, because they are included in the transaction verifier)
- the address book (which is included in the peer set)
- the optional RPC service - did we want to include it, or does that complicate the diagram too much?

I'm mainly working off the list of tasks and services here:
https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/commands/start.rs

Reply:
> I think one of the dependencies might be inverted:
> 
> * all the arrows from the inbound service should go in the same direction, because nothing depends on it

It seems bidirectional, there's a task handle in `PeerSet` for creating new connections that uses inbound for some requests. Inbound is missing an arrow from the state too.

> Optional: I think it might be easier to read if we:
> 
> * switch how dependencies are shown, so the called service is the target of the arrow
> * split the block and checkpoint verifier out of the chain verifier (they have different dependencies, the checkpoint verifier doesn't verify transactions)
> 
> What do you think?

I like that idea. 

> the optional RPC service - did we want to include it, or does that complicate the diagram too much?

I think we should the include the RPC server.

How's this: 

(dotted lines are getblocktemplate-rpcs feature)

![service dependencies](https://user-images.githubusercontent.com/5491350/210673316-4176c91c-451f-49ff-a86f-5b9b22d60deb.svg)

[rendered online](https://dreampuf.github.io/GraphvizOnline/#digraph%20services%20%7B%0D%0A%20%20%20%20transaction_verifier%20-%3E%20state%0D%0A%20%20%20%20mempool%20-%3E%20state%0D%0A%20%20%20%20chain_verifier%20-%3E%20state%0D%0A%20%20%20%20inbound%20-%3E%20state%0D%0A%20%20%20%20rpc_server%20-%3E%20state%0D%0A%20%20%20%20mempool%20-%3E%20transaction_verifier%0D%0A%20%20%20%20chain_verifier%20-%3E%20checkpoint_verifier%0D%0A%20%20%20%20inbound%20-%3E%20mempool%0D%0A%20%20%20%20rpc_server%20-%3E%20mempool%0D%0A%20%20%20%20inbound%20-%3E%20chain_verifier%0D%0A%20%20%20%20syncer%20-%3E%20chain_verifier%0D%0A%20%20%20%20rpc_server%20-%3E%20chain_verifier%20%5Bstyle%3Ddotted%5D%0D%0A%20%20%20%20rpc_server%20-%3E%20peer_set%20%5Bstyle%3Ddotted%5D%0D%0A%20%20%20%20peer_set%20-%3E%20inbound%0D%0A%20%20%20%20syncer%20-%3E%20peer_set%0D%0A%20%20%20%20inbound%20-%3E%20peer_set%0D%0A%20%20%20%20mempool%20-%3E%20peer_set%0D%0A%20%20%20%20rpc_server%20-%3E%20syncer%20%5Bstyle%3Ddotted%5D%0D%0A%20%20%20%20block_verifier%20-%3E%20state%0D%0A%20%20%20%20checkpoint_verifier%20-%3E%20state%0D%0A%20%20%20%20block_verifier%20-%3E%20transaction_verifier%0D%0A%20%20%20%20chain_verifier%20-%3E%20block_verifier%0D%0A%7D)

Reply:
That looks good, but it seems a bit busy. I think it would be good to have in a technical details section.

What do you think of this more conceptual diagram, where I tried to focus on the major data flows within the application?
I deleted some of the dependencies that are used for status information, and I split the inbound and outbound network communication based on which node initiates it. (We could call it local requests and remote requests if that's clearer?)

![graphviz](https://user-images.githubusercontent.com/8951843/210697872-1c610a9c-ade2-4732-bed7-1d6a02304aa5.svg)

[rendered online](https://dreampuf.github.io/GraphvizOnline/#digraph%20services%20%7B%0D%0A%20%20%20%20transaction_verifier%20-%3E%20state%0D%0A%20%20%20%20mempool%20-%3E%20state%0D%0A%20%20%20%20inbound_network%20-%3E%20state%0D%0A%20%20%20%20rpc_server%20-%3E%20state%0D%0A%20%20%20%20mempool%20-%3E%20transaction_verifier%0D%0A%20%20%20%20chain_verifier%20-%3E%20checkpoint_verifier%0D%0A%20%20%20%20inbound_network%20-%3E%20mempool%0D%0A%20%20%20%20rpc_server%20-%3E%20mempool%0D%0A%20%20%20%20inbound_network%20-%3E%20chain_verifier%0D%0A%20%20%20%20syncer%20-%3E%20chain_verifier%0D%0A%20%20%20%20rpc_server%20-%3E%20chain_verifier%20%5Bstyle%3Ddotted%5D%0D%0A%20%20%20%20syncer%20-%3E%20outbound_network%0D%0A%20%20%20%20mempool%20-%3E%20outbound_network%0D%0A%20%20%20%20block_verifier%20-%3E%20state%0D%0A%20%20%20%20checkpoint_verifier%20-%3E%20state%0D%0A%20%20%20%20block_verifier%20-%3E%20transaction_verifier%0D%0A%20%20%20%20chain_verifier%20-%3E%20block_verifier%0D%0A%7D)

Reply:
> That looks good, but it seems a bit busy. I think it would be good to have in a technical details section.
> 
> What do you think of this more conceptual diagram, where I tried to focus on the major data flows within the application? I deleted some of the dependencies that are used for status information, and I split the inbound and outbound network communication based on which node initiates it. (We could call it local requests and remote requests if that's clearer?)

I like the simplification, I added an invisible edge from rpc_server to inbound for layout, and kept the 'peer_set' and 'inbound' node names [here](https://dreampuf.github.io/GraphvizOnline/#digraph%20services%20%7B%0D%0A%20%20%20%20transaction_verifier%20-%3E%20state%0D%0A%20%20%20%20mempool%20-%3E%20state%0D%0A%20%20%20%20inbound%20-%3E%20state%0D%0A%20%20%20%20rpc_server%20-%3E%20state%0D%0A%20%20%20%20mempool%20-%3E%20transaction_verifier%0D%0A%20%20%20%20chain_verifier%20-%3E%20checkpoint_verifier%0D%0A%20%20%20%20inbound%20-%3E%20mempool%0D%0A%20%20%20%20rpc_server%20-%3E%20mempool%0D%0A%20%20%20%20inbound%20-%3E%20chain_verifier%0D%0A%20%20%20%20syncer%20-%3E%20chain_verifier%0D%0A%20%20%20%20rpc_server%20-%3E%20chain_verifier%20%5Bstyle%3Ddotted%5D%0D%0A%20%20%20%20syncer%20-%3E%20peer_set%0D%0A%20%20%20%20mempool%20-%3E%20peer_set%0D%0A%20%20%20%20block_verifier%20-%3E%20state%0D%0A%20%20%20%20checkpoint_verifier%20-%3E%20state%0D%0A%20%20%20%20block_verifier%20-%3E%20transaction_verifier%0D%0A%20%20%20%20chain_verifier%20-%3E%20block_verifier%0D%0A%20%20%20%20rpc_server%20-%3E%20inbound%20%5Bstyle%3Dinvis%5D%20%2F%2F%20for%20layout%20of%20the%20diagram%0D%0A%7D).

Updated with the new svg and links from each node its docs in [`f601bf3`](https://github.com/ZcashFoundation/zebra/pull/5892/commits/f601bf39b515a6ecb0cd26286b6cdfac193cd444).

Reply:
Looks great!

For the blog post, we might want to call the network dependencies:
- requests from peers
- requests to peers

Or we could say that in the text of the post.

Feel free to resolve this comment once you've read it.